### PR TITLE
[kernel] avoid dynamic side effect handling

### LINKED ISSUE #13324: coqchk -o incorrectly reports abstracted subproofs from the obligation tactic in Program Definition as axioms in master
#### Description of the problem

```
$ cat abs.v
```
```coq
Require Import Coq.Program.Basics Coq.Program.Tactics.
Obligation Tactic := abstract exact I.
Program Definition foo : True := _.
Print Assumptions foo.
```
```
$ for ver in master 8.12.0; do (true | ~/.local64/coq/coq-$ver/bin/coqtop -q); ~/.local64/coq/coq-$ver/bin/coqc -q abs.v; ~/.local64/coq/coq-$ver/bin/coqchk -silent -o abs.vo; done
Welcome to Coq jgross-Leopard-WS:/home/jgross/Downloads/coq/coq-master,master (cdfe69d6da6b32338ba74c9f599c74389089c9dd)
Skipping rcfile loading.

Coq <
Closed under the global context

CONTEXT SUMMARY
===============

* Theory: Set is predicative

* Axioms:
    abs.temporary_proof1_subproof

* Constants/Inductives relying on type-in-type: <none>

* Constants/Inductives relying on unsafe (co)fixpoints: <none>

* Inductives whose positivity is assumed: <none>

Welcome to Coq 8.12.0 (August 2020)
Skipping rcfile loading.

Coq <
Closed under the global context

CONTEXT SUMMARY
===============

* Theory: Set is predicative

* Axioms: <none>

* Constants/Inductives relying on type-in-type: <none>

* Constants/Inductives relying on unsafe (co)fixpoints: <none>

* Inductives whose positivity is assumed: <none>

```

This is a regression over 8.12.0.  Interestingly, it only occurs when the `abstract` is in the obligation tactic.

#### Coq Version

cdfe69d6da6b32338ba74c9f599c74389089c9dd

### PR DESCRIPTION
**Pull request by gares**
This PR clean up some code path in the kernel, ensuring side effects are either there or not there (which was previously checked dynamically).

Fix #13324

- https://github.com/maximedenes/vscoq/pull/10

### CODE REVIEW

File: test-suite/output/abstract_export_seff.out
Comment (gares) [inline test-suite/output/abstract_export_seff.out:8]: Before this patch, the various `toto` were printed as axioms (but for the one at line 6).

File: vernac/declare.ml
Comment (SkySkimmer) [inline vernac/declare.ml:775]: reminder to remove these comments before merging

File: vernac/declare.mli
Comment (gares) [inline vernac/declare.mli:293]: since these APIs are specific for the stm, I did rename them so that you know it (and don't call them) if you are working on a non-stm coq toplevel

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:1769]: Since the stm rewinds the state before Qed/Defined, side effects don't leak

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:1708]:
```suggestion
```

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:1631]:
```suggestion
```

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:775]:
```suggestion
```

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:391]:
```suggestion
```

File: test-suite/output/abstract_export_seff.v
Comment (gares) [inline test-suite/output/abstract_export_seff.v:21]:
```suggestion
```

File: test-suite/output/abstract_export_seff.out
Comment (SkySkimmer) [reply test-suite/output/abstract_export_seff.out:8]: Does that mean this PR fixes https://github.com/coq/coq/issues/13324 ?

Comment (gares) [reply test-suite/output/abstract_export_seff.out:8]: tested locally, yes

File: vernac/declare.mli
Comment (SkySkimmer) [reply vernac/declare.mli:293]: shouldn't you then also rename the ones in vernacstate?

File: vernac/declare.ml
Comment (SkySkimmer) [reply vernac/declare.ml:1765]: this seems like a comment that should be in the code

File: kernel/safe_typing.ml
Comment (SkySkimmer) [inline kernel/safe_typing.ml:997]: This essentially reverts https://github.com/coq/coq/pull/12935, I guess that's why this PR fixes #13324

**Review by SkySkimmer — APPROVED** [review]
@coqbot bench

**Comment by SkySkimmer** [comment]
test suite fails

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:997]: Beware that while #12935 caused the observable bug, as commented in #13324 the previous state was incorrect in a different way, i.e. gladly keeping around dead subproofs. We should check what the current does exactly.

**Comment by gares** [comment]
yes, I noticed but no time to fix it yet

**Comment by gares** [comment]
@coqbot minimize ci-fiat_parsers

**Comment by gares** [comment]
@coqbot ci minimize ci-fiat_parsers

File: kernel/safe_typing.mli
Comment (ejgallego) [inline kernel/safe_typing.mli:91]: Can we document this function?

Comment (ejgallego) [inline kernel/safe_typing.mli:91]: Can we document this function?

File: vernac/declare.ml
Comment (ejgallego) [inline vernac/declare.ml:110]: Can we document this? What's does "leaking side effect" mean?

File: vernac/declare.ml
Comment (ejgallego) [inline vernac/declare.ml:2104]: This new parameter seems strange, proof_info already contains an ending, why the need to pass another one?

File: vernac/declare.mli
Comment (ejgallego) [reply vernac/declare.mli:293]: Why are they specific to the stm ? What should a non-stm toplevel call instead?

[instead of the renaming you can also just wrap them in a `Stm` module for namespacing]

File: vernac/declare.ml
Comment (ejgallego) [inline vernac/declare.ml:2168]: Same comment about the extra `ending` parameter where we are already passing `pinfo` to the functions.

**Comment by gares** [comment]
I'll document properly things when they will work.

 A proof that leaks side effects (eg abstract) is the norm, unless the DM decides to fiddle with states, as the STM does, and rewind over the effects before declaring the proof. If you delegate to a worker you are kind of obliged to hide side effects.
The code paths need to know the intention, indeed only the stm_* entry points set it to false.

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml:2168]: right, but it seems out of place since some choices that regard the ending need to be taken before that time. In particular I don't need the full ending info, just if is the equation/deriving case, which are ugly hacks, see where I use this extra ending.

Anyway, I can rectify that.

File: kernel/safe_typing.mli
Comment (gares) [reply kernel/safe_typing.mli:91]: sure, the old entry point was tolerating the two cases dynamically, now you have two entries, one if the effects are in the env, one if they are not.

**Comment by gares** [comment]
Thanks @coqbot!

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml]: done

**Comment by gares** [comment]
Still a but with equations.
@coqbot ci minimize ci-metacoq

**Comment by gares** [comment]
@coqbot ci minimize ci-metacoq

**Comment by JasonGross** [comment]
It looks like coqbot encountered a separate anomaly in the middle of minimizing (perhaps the log is enough to recreate it? if not, once the minimizer is finished, you can download the verbose log and see the file that generated this error, and even ask the minimizer to minimize this error with `@coqbot resume ci minimize ci-metacoq` followed by the contents of the file in a triple-backquoted code block)
```
I will now attempt to admit definitions with Admitted
Non-fatal error: Failed to admit definitions and preserve the error.  
The new error was:
File "/tmp/tmpkttso8w0/MetaCoq/Erasure/EInduction.v", line 1993, characters 0-9:
Error: Anomaly "in Lemmas.save_lemma_admitted: more than one statement."
Please report at [http://coq.inria.fr/bugs/.](http://coq.inria.fr/bugs/)
Intermediate code not saved.
```

**Comment by JasonGross** [comment]
Resumption failed because
```
git checkout 4a9b926e4283e50bf6e5d95b6d747854986833a5
  fatal: reference is not a tree: 4a9b926e4283e50bf6e5d95b6d747854986833a5
```
And unfortunately the example is just a bit too long to resume minimization by comment

**Comment by JasonGross** [comment]
Let's just ask again @coqbot ci minimize ci-metacoq

**Review by ejgallego — COMMENTED** [review]
Umm, that is indeed tricky, not sure we can do with this API, or if we will need a kind of two-phase `Qed` setup, where `Declare.Proof.start` indeed takes a parameter and `save` can does return something like `type t = Saved | ToExport of env * effs`.

I recall the disucssion at https://github.com/coq/coq/wiki/BreakOut-2020-12-01-envs which I think is related.

**Comment by JasonGross** [comment]
MetaCoq inlining failure due to https://github.com/JasonGross/coq-tools/issues/129.  I can work around it for MetaCoq, but in general this is hitting the problem that there's no way to create a module named Prop.Equations from within a single file, because Prop is a keyword.

**Comment by gares** [comment]
@coqbot bench

File: vernac/declare.mli
Comment (gares) [reply vernac/declare.mli:293]: This whole module is deprecated, I believe you did deprecate it.
IMO a regular toplevel should just vernacinterp the Qed or Defined, and not fiddle with proofs directly.

**Comment by gares** [comment]
If there are no surprises in the bench, I believe this PR is ready. It removes from the kernel some dynamic dispatching, enforcing more discipline in the upper layers about effects, which are no more duplicate blindly when multiple constants are declared at once. It also inhabits (export) private constants ASAP so that they are not considered as axioms.

FTR I've also tested this commit in #16361 and it works as expected.

File: test-suite/output/abstract_export_seff.v
Comment (silene) [inline test-suite/output/abstract_export_seff.v:20]: Why no `Print toto4.`?

Also, should we add the following?
```coq
Definition foo5 : nat.
Proof.
    abstract (exact 1) using toto5.
Abort.
Fail Print toto5.
```

**Comment by JasonGross** [comment]
Should the examples that coqbot minimized be added to the test-suite?
(Also, btw, minimization of metacoq should work fine once https://github.com/JasonGross/coq-tools/pull/133 is merged)

File: kernel/safe_typing.ml
Comment (JasonGross) [inline kernel/safe_typing.ml:791]:
```suggestion
    CErrors.anomaly Pp.(str"export_side_effects: already existing seff");
```

File: kernel/safe_typing.ml
Comment (JasonGross) [inline kernel/safe_typing.ml:838]:
```suggestion
    CErrors.anomaly Pp.(str"export_existing_side_effects: nonexistent seff");
```

File: test-suite/output/abstract_export_seff.v
Comment (gares) [reply test-suite/output/abstract_export_seff.v:20]: In this branch, or master, `Print toto4` should fail. Not because the kernel or vernac, but because the STM wants to do so, that is make proof of `foo4` be the same weather it is delegated or not. It is easy to make `toto4` appear in the env if the proof is run locally, impossible if delegated (since the whole point of delegating is to let master continue, and one has to know the env in order to continue, and statically you can't tell if abstract will be executed or not in general, and you don't even know the type of the extra name...).

In my first version of the test I did put a `Fail Print toto4` but I removed it, since this is again an STM specific behavior. A simple compiler like #16361 that just fold interp makes toto4 visible (as of today, it can be changed of course).

I think foo5 is in a similar situation. It works today with the STM but would not work by just folding interp (without more changes).

So, I'm adding the tests, but I believe they should not be "the spec" of the language. I mean a Coq top level should be free to let effects escape Qed or Abort, or not and inline/erase them. It is a bit like undefined behavior to me.

File: vernac/declare.ml
Comment (ejgallego) [inline vernac/declare.ml:112]: Maybe this could use some rephrasing, I'm not sure I understand it. Is that equivalent to:

"The proof is going to be typechecked in an enviroment where the side effects of the proof are present" ?

Comment (gares) [reply vernac/declare.ml:112]: "...is declared in an environment which does *not* contain its side effects..."

Comment (gares) [reply vernac/declare.ml:112]: they are not there yet, it is up the declaration code to export them

**Comment by gares** [comment]
> Should the example from metacoq be added to the test-suite (is there a way to make an example that does not depend on Equations?) (Also, btw, minimization of metacoq should work fine once [JasonGross/coq-tools#133](https://github.com/JasonGross/coq-tools/pull/133) is merged)

The file you gave me was too long, depending on Equations an Metacoq-template. I did not try to minimize it by hand.

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml:112]: Maybe "proof_construction_leaks_effects" (default to true, false when the STM is involved)

Comment (ejgallego) [reply vernac/declare.ml:112]: I see, thanks. What was the reason the STM did rewind the environment?

**Review by ejgallego — COMMENTED** [review]
This stuff is very tricky in general, I'd be nice if we could document what the invariants for `Declare.Proof.save_proof` are, in terms of environment.

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml:112]: Because, back then, I wanted the code paths of Qed and Defined to be "the same" and Qed needs to rewind since it has to hide the effects (by inlining them). See the message I wrote to Guillaume about the test to refresh your memory on why.

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml:394]: To my understanding yes, see the addition to declare_private_constant

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml:1946]: You mean `prepare_proof`? I think I had to do it here because in the case of mutual stuff we have only one proof entry which is declared many times (I was surprised by that). The "multiplicity" is dictated by the `pinfo.Proof_info.cinfo` see the last line of declare_mutdef.

File: vernac/declare.ml
Comment (ejgallego) [reply vernac/declare.ml:112]: I see, so does the linear compiler still need to rewind for `Qed` ?

Comment (gares) [reply vernac/declare.ml:112]: I tried a quick commit to see if I could change that (in the STM) but it did not work out of the box, so I decided to leave this PR as is. It is already large enough imo.

File: vernac/declare.ml
Comment (ejgallego) [reply vernac/declare.ml:1946]: Yes, that's a big problem which is not new to this PR, maybe we should move the comment you remove in the place this was taken care before and add it here?

Because this sees like a fixme to me.

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml:112]: Well, if you want to hide the effects yes. But I don't think you need to. This is why I added the extra tests suggested by Guillaume as a separate commit mentioning the STM (see also my answer to his message).

Comment (ejgallego) [reply vernac/declare.ml:112]: Indeed maybe it is just easier to add a table for `Declare.Proof.save` that specs the pre-condition on the caller w.r.t. envs, so at least, we have a spec somewhere (or a desired spec).

Comment (gares) [reply vernac/declare.ml:112]: Once could also have an API to pop from the env a constant under some circumstances, and hide effects that way. My previous answer was considering the APIs we have today.

**Comment by gares** [comment]
> IMO ideally side effects would be in the evar map but not the global env until the end of the proof. This would also avoid errors like

That would be OK to me, but many things would become odd. In particular the proof term being build would not type check trivially (I mean a tactic calls abstract, then typechecks the proof term, boom). If it was simpler, I would have probably opted for that back when I wrote the STM.

**Comment by gares** [comment]
> (such errors can happen in the wild typically when rewrite auto generates schemes but then fails to rewrite, I don't remember the issue number right now)

To summarize, this PR does not solve this issue.

File: vernac/declare.ml
Comment (ejgallego) [reply vernac/declare.ml:112]: Ok thanks, so if I understand correctly, there is some choice w.r.t. what the callers can do w.r.t. the env, that's fine, but I'd be great if we can document it.

Indeed we don't need to do more stuff in this PR, but I feel that writing down the current invariants could be very helpful.

File: vernac/declare.ml
Comment (gares) [reply vernac/declare.ml:1946]: What do you want me to write? The command was saying that one should not duplicate side effects and that maybe the right thing to do is to declare them once with the first constant. This is what the patch does, no? Bringing it back seems odd, since that issue is solved.

Cleaning up to the code so that we generate one proof entry per copy of the mutfix is OK to me, but maybe in another PR?

File: test-suite/output/abstract_export_seff.v
Comment (silene) [reply test-suite/output/abstract_export_seff.v:20]: I see. I don't quite agree with your conclusion though. If we consider that it should be possible to delegate any `Qed`-ended proof (I agree), then we should strive as much as possible to make it **as if** it is always delegated. We cannot have *undefined behaviors* in a proof assistant. The simple fact of saying so is bound to come back and bite us. At the very least, the documentation could be changed from "If the proof is closed with Qed, the auxiliary lemma is inlined in the final proof term" to "If the proof is closed with Qed, the auxiliary lemma is not guaranteed to appear in the environment". (And the warning about asynchronous proofs should be removed or replaced by this sentence.)

**Comment by ppedrot** [comment]
@SkySkimmer Currently you can't remove private constants from the environment because you need them in the nametab. There is the infamous `abstract foo using bar` variant that allows giving a reliable name to the constant. I tried several time working around this, e.g. replacing the global constant with a proof-local let-in so that most of the calls to the nametab are avoided, but some developments (fiat-crypto IIRC) rely critically on the constant existing globally because the calls are not well-scoped.

There is work to do but it's not an easy one.

**Comment by gares** [comment]
The simple compiler branch raised a bug (which cannot be reached if you use the STM), but I did push the fix here anyway.

**Comment by gares** [comment]
So that fix was OK.

**Review by ppedrot — APPROVED** [review]
The kernel changes look OK to me (although this is very tricky code). I can't vouch for the declare part, since this is even trickier.

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:997]: I checked the vo file size on the bench, and it didn't seem to vary that much. Something may have happened in the meantime that solved the underlying bug?

**Comment by ejgallego** [comment]
> The kernel changes look OK to me (although this is very tricky code). I can't vouch for the declare part, since this is even trickier.

It is hard to review this code until the intended semantics / preconditions are documented in the `.mli` for the affected calls.

File: vernac/declare.ml
Comment (ejgallego) [reply vernac/declare.ml:1946]: I don't know what to write as I don't understand the current code generating this now.

File: vernac/declare.ml
Comment (ejgallego) [inline vernac/declare.ml:1648]: This needs a comment, I can't figure out what is going on here.

**Review by ejgallego — CHANGES_REQUESTED** [review]
I think PR looks good however there is not way for me to understand if the code meets the intended spec with the current documentation. Thus it needs two pieces of documentation:
- `declare.mli`: State what contracts do the callers to `Declare.Proof.save` must obey w.r.t. the env.
- `declare.ml`: Document what is going on now in `prepare_proof`

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:1649]:
```suggestion
    (* When a bunch of constants are generated from the same proof session effects;
        - are attached to the first one if the constants are transparent
        - are attached to each constants if the constants are opaque
       This is because effects are not inlined in transparent constants (so the effects
       of the first constant are also visible by the other ones), while they are in case
       of opaque constants, so they are hidden inside each of them.
       As an optimization one could attach only the ones actually used in the
       opaque case. *)
    let fst_eff, more_eff =
```

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:1659]:
```suggestion
  (* derive skips the first goal *) 
  | Proof_ending.End_derive _ -> regular_preparation (List.tl initial_goals)
```

File: vernac/declare.ml
Comment (gares) [inline vernac/declare.ml:1660]:
```suggestion
  (* Equations calls prepare_proof even if there is no goal *)
  | Proof_ending.End_equations _ when initial_goals = [] ->
```

File: vernac/declare.mli
Comment (gares) [inline vernac/declare.mli:292]:
```suggestion
  (* The STM rewinds the system state between closing the proof and saving it, both in the transparent
      and opaque case. These APIs expect the caller to do so. If proof construction generated side effects, then
      not respecting this contract results in assertion (from the kernel) that the side effects are already present *)
  val stm_close_proof : opaque:Vernacexpr.opacity_flag -> keep_body_ucst_separate:bool -> t -> proof_object
```

**Comment by gares** [comment]
Here the comments, I hope I understood what you were asking

**Comment by gares** [comment]
ping

**Comment by ejgallego** [comment]
A bit short of time for the next couples of days but this is in my queue.

**Comment by gares** [comment]
I think we should review our processes, this PR has spent half of its life idle.

File: vernac/declare.ml
Comment (ejgallego) [reply vernac/declare.ml:1649]: That reads great thanks!

**Comment by ejgallego** [comment]
Hi @gares , the comments look great, thanks. Only thing I'm not sure yet what the intended contract for `Declare.Proof.save` is w.r.t. env. That's the main path this PR updates for the simple compiler.

Can we document it ?

**Comment by gares** [comment]
My understanding is that that code path did not change per se, that is the simple compiler calls it in the final proof env where side effects are there.
What changed is that these effects are "exported", that means they are given the regular status of proved constants. Indeed I did fix #13324 (see also the test output here) for both the simple compiler and the stm. In the stm case side effects were (and still are) exported at Defined time, but were not exported in the proof construction, so `abstract t. Print foo_subproof` would show an axiom (so the bad behavior was kind of transient).
Since the simple compiler does not follow the code paths I did signal with the `stm_` prefix, effects were not exported ad Defined time, so they were always displayed as axioms. This was actually why I started writing this PR (see the branch name), and while doing so I figured that by fixing that I could also make the kernel invariants more strict, hence the PR title.

What do you want me to write in the code? IMO I did not change how save works.

**Comment by ejgallego** [comment]
Yeah I was trying to understand how/if the code path in `Proof.save` had changed.

I think it'd be nice if we write down what env `Proof.save` does expect, in both cases, something like:

```
val save : ...
(** [save pm proof opaque idtop] typecheck a proof, and register it with the global environment. The caller must ensure that the global environnent is in shape X when `opaque = Defined` and in shape Y when `opaque = Qed`
```

Even better would be to have this file not use `Global.env`, but that's a lot of work.

**Comment by gares** [comment]
Done, but as I said this doc has little to do with this PR IMO.

**Comment by ejgallego** [comment]
> Done, but as I said this doc has little to do with this PR IMO.

Thanks Enrico. That part about this PR not having to do with that contract is confusing to me tho.

Let me recap to see if I understand what is going on:
- This PR is needed to improve the behavior of the "simple compiler"
- The simple compiler calls `Declare.Proof.save`
- So this PR alters the behavior of `Declare.Proof.save` , right?

File: vernac/declare.mli
Comment (ejgallego) [inline vernac/declare.mli:204]: Ok, so both for `Defined` and `Qed` the input env has to be "linear" one.

But we still have cases in the CI for the simple compiler where this is not working properly, right?

For example, https://github.com/coq/coq/pull/16361/checks?check_run_id=7774483046 is with this PR applied right?

Comment (gares) [reply vernac/declare.mli:204]: This PR repairs some cases which are tested here and in the simple compiler about side effects being displayed as axioms. Some of these cases are also visible without the simple compiler, hence this PR is closing an existing issue.

This PR does not solve all problems we have with the simple compiler, nor claims of doing so.

In passing is simplifies the kernel.

Comment (ejgallego) [reply vernac/declare.mli:204]: Ok, so indeed we don't yet met the contract, but at least we now have a intention about what the contract should be :+1:

**Review by ejgallego — APPROVED** [review]
Looks good to me, tho unfortunately still the contract we expose in the main API is not respected by our implementation.

The API for declaration is indeed still tricky, there are some ideas around in different places  (wiki, issues) resulting from discussions, but I think that a deeper joint design discussion will be needed to fully fix it.

**Comment by gares** [comment]
The async test suite reveals a discrepancy for admitted proofs containing side effects (IIRC the STM has a special case). I'm adding a silly commit to have the full picture on CI.

**Comment by ppedrot** [comment]
What's the status of this?

**Comment by gares** [comment]
Lack of time to finalize it.
The only problem, iirc, is that due to different code paths in the STM, one test showed different behavior in async mode.

**Pull request closed without merging**
